src/audio/voice/capture/voice_capture.cpp

```diff
@@ -0,0 +1,165 @@
+#include "voice_capture.hpp"
+#include "constants.hpp"
+
+namespace AbaddonClient::Audio::Voice {
+
+using OpusEncoder = Opus::OpusEncoder;
+using EncoderSettings = OpusEncoder::EncoderSettings;
+using SignalHint = OpusEncoder::SignalHint;
+using EncodingApplication = OpusEncoder::EncodingApplication;
+
+void capture_callback(ma_device* pDevice, void* pOutput, const void* pInput, ma_uint32 frameCount) {
+    auto capture = reinterpret_cast<VoiceCapture*>(pDevice->pUserData);
+    if (capture == nullptr) {
+        return;
+    }
+
+    const auto buffer = InputBuffer(static_cast<const float*>(pInput), CAPTURE_BUFFER_SIZE);
+    capture->OnAudioCapture(buffer);
+}
+
+VoiceCapture::VoiceCapture(Context &context) noexcept :
+    m_device(context, GetDeviceConfig(), context.GetActiveCaptureID()) {}
+
+void VoiceCapture::Start() noexcept {
+    StartEncoder();
+    if (!m_encoder.Lock()->has_value()) {
+        return;
+    }
+
+    m_device.Start();
+}
+
+void VoiceCapture::Stop() noexcept {
+    m_device.Stop();
+    m_encoder.Lock()->reset();
+}
+
+void VoiceCapture::SetActive(bool active) noexcept {
+    m_active = active;
+}
+
+void VoiceCapture::SetCaptureDevice(ma_device_id &&device_id) noexcept {
+    spdlog::get("voice")->info("Setting capture device");
+
+    const auto success = m_device.ChangeDevice(std::move(device_id));
+    if (!success) {
+        spdlog::get("voice")->error("Failed to set capture device");
+    }
+}
+
+ma_device_config VoiceCapture::GetDeviceConfig() noexcept {
+    auto config = ma_device_config_init(ma_device_type_capture);
+    config.capture.format = CAPTURE_FORMAT;
+    config.sampleRate = CAPTURE_SAMPLE_RATE;
+    config.capture.channels = CAPTURE_CHANNELS;
+    config.pulse.pStreamNameCapture = "Abaddon (Capture)";
+
+    config.periodSizeInFrames = CAPTURE_FRAME_SIZE;
+    config.performanceProfile = ma_performance_profile_low_latency;
+
+    config.dataCallback = capture_callback;
+    config.pUserData = this;
+
+    return config;
+}
+
+void VoiceCapture::StartEncoder() noexcept {
+    auto settings = EncoderSettings {};
+    settings.sample_rate = CAPTURE_SAMPLE_RATE;
+    settings.channels = CAPTURE_CHANNELS;
+    settings.bitrate = 64000;
+    settings.signal_hint = SignalHint::Auto;
+    settings.application = EncodingApplication::Audio;
+
+    auto encoder = OpusEncoder::Create(settings);
+    if (!encoder) {
+        return;
+    }
+
+    m_encoder.Lock()->emplace(std::move(*encoder));
+}
+
+void VoiceCapture::OnAudioCapture(InputBuffer input) noexcept {
+    if (!m_active) {
+        return;
+    }
+
+    CaptureBuffer buffer;
+    std::copy(input.begin(), input.end(), buffer.begin());
+
+    ApplyEffects(buffer);
+    if (ApplyNoise(buffer)) {
+        EncodeAndSend(std::move(buffer));
```

Comment:
I suppose taking `CaptureBuffer&` would be better to avoid copy, I was trying to use rvalue reference to express ownership, but looks like that just copies `std::array`

Reply:
yeah `CaptureBuffer&` is totally fine. array is basically just a wrapper over a c-style array so theres nothing for a move to steal from